sjsonnet/src/sjsonnet/Std.scala

```diff
@@ -1141,68 +1056,64 @@ class Std {
+        case str: Val.Str if !settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
       val b = args(1) match {
         case arr: Val.Arr => arr.asLazyArray
-        case str: Val.Str => stringChars(pos, str.value).asLazyArray
+        case str: Val.Str if !settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
         case _ => Error.fail("Arguments must be either arrays or strings")
       }
 
       val keyF = args(2)
       val out = new mutable.ArrayBuffer[Lazy]
 
       for (v <- a) {
-        if (keyF.isInstanceOf[Val.False]) {
-          val vf = v.force
-          if (!b.exists(value => {
-            ev.equal(value.force, vf)
-          }) && !out.exists(value => {
-            ev.equal(value.force, vf)
-          })) {
-            out.append(v)
-          }
-        } else {
-          val keyFFunc = keyF.asInstanceOf[Val.Func]
-          val appliedX = keyFFunc.apply1(v, pos.noOffset)(ev)
-
-          if (!b.exists(value => {
-            val appliedValue = keyFFunc.apply1(value, pos.noOffset)(ev)
-            ev.equal(appliedValue, appliedX)
-          }) && !out.exists(value => {
-            val mValue = keyFFunc.apply1(value, pos.noOffset)(ev)
-            ev.equal(mValue, appliedX)
-          })) {
-            out.append(v)
-          }
+        if (existsInSet(ev, pos, keyF, b, v.force) && !existsInSet(ev, pos, keyF, out.toArray, v.force)) {
+          out.append(v)
         }
       }
-
-      sortArr(pos, ev, new Val.Arr(pos, out.toArray), keyF)
+      if (settings.strictSetOperations) {
+        new Val.Arr(pos, out.toArray)
+      } else {
+        sortArr(pos, ev, new Val.Arr(pos, out.toArray), keyF)
+      }
     },
-    builtinWithDefaults("setMember", "x" -> null, "arr" -> null, "keyF" -> Val.False(dummyPos)) { (args, pos, ev) =>
+    builtinWithDefaults("setDiff", "a" -> null, "b" -> null, "keyF" -> Val.False(dummyPos)) { (args, pos, ev) =>
+      val a = args(0) match {
+        case arr: Val.Arr => arr.asLazyArray
+        case str: Val.Str if !settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
+        case _ => Error.fail("Arguments must be either arrays or strings")
+      }
+      val b = args(1) match {
+        case arr: Val.Arr => arr.asLazyArray
+        case str: Val.Str if !settings.strictSetOperations => stringChars(pos, str.value).asLazyArray
+        case _ => Error.fail("Arguments must be either arrays or strings")
+      }
+
       val keyF = args(2)
+      val out = new mutable.ArrayBuffer[Lazy]
+
+      for (v <- a) {
+        if (!existsInSet(ev, pos, keyF, b, v.force) && !existsInSet(ev, pos, keyF, out.toArray, v.force)) {
+          out.append(v)
+        }
+      }
 
-      if (keyF.isInstanceOf[Val.False]) {
-        val ujson.Arr(mArr) = Materializer(args(1))(ev)
-        val mx = Materializer(args(0))(ev)
```

Comment:
FWIW, I've found that removing the materializer here is an _enormous_ performance improver for some inputs. In one real-world case I saw a whole _third_ of all memory allocations occurring from this materializer call.

I noticed that a few other built-in functions have similar use of `Materializer` and I'm wondering whether there's a principled reason for that or whether it's safe for us to simply remove all non-output / non `toString` uses of `Materializer`.